### ISSUE TITLE
Add deferred decoding support for Cadence values

### ISSUE THREAD

**Issue by SupunS** [onflow/cadence#844]
### Issue To Be Solved

When the cadence values are read from the storage, all the raw values are decoded to go-types. However, sometimes not all these values are used. e.g: 
- In composite types, some fields may never be used.
- In array types, not all members be accessed.

So always decoding all the values is a bit of extra work, and could be a waste of resources if they are never used.

### Suggest A Solution
The idea is to delay the decoding until a value actually used.

This could be done by:
- When encoding a value, write an additional property `content-length`, before writing the content.
  e.g:` <type-tag>  <content-length>  <content>`
- When reading back from the storage, read the `content-length` and load the raw-bytes for the `content`. Do not try to decode.
- So the cadence-value will internally hold the raw bytes.
- Decode the raw bytes on-demand, if the values  actually used in some way (e.g: access a field, logging,  etc.). This again is a shallow loading. Only loaded up to a one-level down.
- Once decoded, cache the decoded values.

This could be extended to:
- [x] Composite values
- [x] Array values
- [ ] Map values

**Comment by SupunS** [comment]
Previously I suggested encoding values in the form of: `<type-tag> <content-length> <content>`

I realized we could further optimize this by breaking the 'content' into two sections such as `meta-content` and `children-content`
- `meta-content` - Contains encoded meta-information about the value. eg.: type-name, kind, location, etc.
- `children-content` - Contains encoded children. eg: fields of a struct, elements of an array, etc.

The reason for the separation is, there are certain operations that require only meta-information about the value. e.g: casting, runtime type checking, etc. In such cases, we could only decode the meta-information, and avoid having to decode children of the collection value.

So the more optimized version of this is:
 ```
<type-tag>  <meta-content-length>  <meta-content>  <children-content-length>  <children-content>
```

**Comment by bluesign** [comment]
for better performance, is it not better to cache this metadata and store type guid?

**Comment by SupunS** [comment]
That's a good thought. It is precisely how it works at the moment. The current meta-data: `type-name` + `location` is combined to produce the guid for the type. The type itself is not stored and only loaded on demand from the runtime.

In this design, I suggested this metadata section not only for type information, but also for any future additions of information, and to make it extensible.

**Comment by bluesign** [comment]
I am trying to understand how random access will work in this model, basically you are loading sequentially but not decoding till value is accessed right? If I have array of 100 items, 100 items have their raw bytes set in this model, not accessed item (let's say 42) only ? then accessed item, decoding only it's fields to raw bytes somehow.

**Comment by SupunS** [comment]
> I am trying to understand how random access will work in this model, basically you are loading sequentially but not decoding till value is accessed right? If I have array of 100 items, 100 items have their raw bytes set in this model, not accessed item (let's say 42) only ? then accessed item, decoding only it's fields to raw bytes somehow.

Yes, that is correct.

**Comment by bluesign** [comment]
if this is the case,  

<type-tag> <children count> [<children offsets> ... ] [<children-content> ... ]

would not be more optimized? 

You could hold data without moving around (with some relative offset calculation) and could have total random access.

**Comment by SupunS** [comment]
Thats interesting!

The encoding format `<type-tag>  <meta-content-length>  <meta-content>  <children-content-length>  <children-content>` is for the high-level container value. Then, depending on the type of the container, we indeed have to encode/decode children (i.e: `<children-content>`) differently. e.g: fields of composites values, elements in arrays, etc.

And of course we could use the format similar to what you suggested for array elements.

**Comment by SupunS** [comment]
Oh, I see your point. That's a really good one actually! Thanks for these great ideas.

I would love to do a perf/memory evaluation and see how the two approaches behave for different use-cases.

**Comment by SupunS** [comment]
Just realize having a single large buffer with position indexes is not optimal/suitable for updates.
e.g: Some optional field originally containing `nil` is updated to have a different value.

**Comment by bluesign** [comment]
@SupunS Actually it should be, but maybe I didn't totally understand the case here. 
Do you also want to optimize encode logic too? 

Considering optimizing the saving part: (If I understood correctly) 

Lets say I have 3 fields(A,B,C), with indexes (0,1,2). Normally I have 3 pointers (offsets) pointing to structure:  (type-tag, children-count, offsets etc). If we set B to some value, we will append the data to large buffer, change pointer to that offset. But when writing data to disk or committing to memory, we will not visit the old offset, so it should automatically compact the space.

**Comment by SupunS** [comment]
But since the old value is not written and space is shrinked, wouldn't that change the indexes? i.e: when reading back values from the disk, the indexes point to a wrong location? Sorry if I'm missing something here.

**Comment by bluesign** [comment]
If you mean offsets by indexes, yes it will change, but when writing you are calculating them. It is basically counter to bytes written. 

Offsets are like header data, so basically, if you know how many fields will be (aka. children count) (let's assume UInt64 for children-count and offsets) So you know where children-content should start.

** HEADER ** DATA 
**<type-tag> <children count> [<children offsets> ... ]** [<children-content> ... ]

For the first field, I am putting offset 0, then I encode child content, write child-content, add bytes-written to current-offset
For the second field, I write current-offset, then I encode child content,  write child-content, add bytes-written to current-offset
...

For each of the child content, actually, we are doing the same thing. They have relative offset references ( so they don't need to even know the main offset, also you can encode children in parallel with this ) 

I hope this is more clear.

**Comment by SupunS** [comment]
Oh, you are suggesting to re-calculate the offsets/indexes. Yup, that would work.

**Comment by SupunS** [comment]
Anyway, all these discussed approaches require the encoding schema to be redesigned in some way. There is a separate effort being made on revisiting the encoding architecture, and these ideas are greatly useful for that effort :open_hands:

The original objective of this issue/epic was to simply delay the decoding until the value's content is accessed without really changing much of the encoding, for the container types. This is now already added for arrays and composite values. We can wrap up this issue, once the same is added for maps as well.

**Comment by bluesign** [comment]
Yeah I already saw results, they look amazing, congratulations
